tests/unattended/unit/framework/bach.sh

```diff
@@ -0,0 +1,642 @@
+    declare -a all_tests
+    mapfile -t all_tests < <(bach-run-tests--get-all-tests)
+    @echo "1..${#all_tests[@]}"
+    for name in "${all_tests[@]}"; do
+        # @debug "Running test: $name"
+        friendly_name="${name/#test-/}"
+        friendly_name="${friendly_name//-/ }"
+        friendly_name="${friendly_name//  / -}"
+        : $(( ++total ))
+        testresult="$(@mktemp)"
+        @set +e
+        assert-execution "$name" &>"$testresult"; test_retval="$?"
+        @set -e
+        if [[ "$name" == test-ASSERT-FAIL-* ]]; then
+            test_retval="$(( test_retval == 0?1:0 ))"
+            test_name_assert_fail="${color_err}ASSERT FAIL${color_end}"
+            friendly_name="${friendly_name/#ASSERT FAIL/}"
+        else
+            test_name_assert_fail=""
+        fi
+        if [[ "$test_retval" -eq 0 ]]; then
+            builtin printf "${color_ok}ok %d - ${test_name_assert_fail}${color_ok}%s${color_end}\n" "$total" "$friendly_name"
+        else
+            : $(( ++error ))
+            builtin printf "${color_err}not ok %d - ${test_name_assert_fail}${color_err}%s${color_end}\n" "$total" "$friendly_name"
+            {
+                builtin printf "\n"
+                @cat "$testresult" >&2
+                builtin printf "\n"
+            } >&2
+        fi
+        @rm "$testresult" &>/dev/null
+    done
+
+    declare color_result="$color_ok"
+    if (( error > 0 )); then
+        color_result="$color_err"
+    fi
+    builtin printf -- "# -----\n#${color_result} All tests: %s, failed: %d, skipped: %d${color_end}\n" \
+           "${#all_tests[@]}" "$error" "$(( ${#all_tests[@]} - total ))">&2
+    [[ "$error" == 0 ]] && [[ "${#all_tests[@]}" -eq "$total" ]]
+}
+
+function bach-on-exit() {
+    if [[ -o xtrace ]]; then
+        exec 8>&2
+        BASH_XTRACEFD=8
+    fi
+    if [[ "$?" -eq 0 ]]; then
+        [[ "${BACH_DISABLED:-false}" == true ]] || bach-run-tests
+    else
+        builtin printf "Bail out! %s\n" "Couldn't initlize tests."
+    fi
+}
+
+trap bach-on-exit EXIT
+
+function @generate_mock_function_name() {
+    declare name="$1"
+    @echo "mock_exec_${name}_$(@dryrun "${@}" | @shasum | @cut -b1-7)"
+}
+export -f @generate_mock_function_name
+
+function @mock() {
+    declare -a param name cmd func body desttype
+    name="$1"
+    if [[ "$name" == @(builtin|declare|eval|set|unset|true|false|read) ]]; then
+        @die "Cannot mock the builtin command: $name"
+    fi
+    if [[ command == "$name" && "$2" != -* ]]; then
+        shift
+        name="$1"
+    fi
+    desttype="$(@type -t "$name" || true)"
+    for param; do
+        if [[ "$param" == '===' ]]; then
+            shift
+            break
+        fi
+        cmd+=("$param")
+    done
+    shift "${#cmd[@]}"
+    if [[ "$name" == /* ]]; then
+        @die "Cannot mock an absolute path: $name"
+    elif [[ "$name" == */* ]] && [[ -e "$name" ]]; then
+        @die "Cannot mock an existed path: $name"
+    fi
+    @debug "@mock $name"
```

Comment:
Same

Reply:
`bach.sh` is the external framework. We will not modify it.